runtime/src/main/java/org/corfudb/runtime/FastSmrMapsLoader.java

```diff
@@ -0,0 +1,319 @@
+        try {
+            logEntry = logData.getLogEntry(runtime);
+        } catch (Exception e) {
+            log.warn("Cannot deserialize log entry" + logData.getGlobalAddress());
+            e.printStackTrace();
+            throw e;
+        }
+
+        return logEntry;
+    }
+
+    /** Extract log entries from logData and update the Corfu Objects
+     *
+     * @param logData LogData received from Corfu server.
+     */
+    private void processLogData(ILogData logData) {
+        LogEntry logEntry;
+        try {
+            logEntry = deserializeLogData(logData);
+        } catch (Exception e) {
+            return;
+        }
+
+        if (logEntry.getType() == LogEntry.LogEntryType.SMR) {
+            // Just one stream, always
+            UUID streamId = logData.getStreams().iterator().next();
+            applySmrEntryToStream(streamId, (SMREntry) logEntry);
+        }
+        else if (logEntry.getType() == LogEntry.LogEntryType.MULTIOBJSMR) {
+            MultiObjectSMREntry multiObjectLogEntry = (MultiObjectSMREntry) logEntry;
+            multiObjectLogEntry.getEntryMap().forEach((stream, multiSmrEntry) -> {
+                multiSmrEntry.getSMRUpdates(stream).forEach((smrEntry) -> {
+                    applySmrEntryToStream(stream, smrEntry);
+                });
+            });
+
+        }
+    }
+
+    private void findAndSetLogHead() {
+        logHead = 0;
+    }
+
+    private void findAndSetLogTail() {
+        logTail = runtime.getSequencerView().nextToken(Collections.emptySet(), 0).getTokenValue();
+    }
+
+    /** Initialize log head and log tails
+     *
+     * If logHead and logTail has not been initialized by
+     * the user, initialize to default.
+     *
+     */
+    private void initializeHeadAndTails() {
+        if (logHead < 0) {
+            findAndSetLogHead();
+        }
+
+        if (logTail < 0) {
+            findAndSetLogTail();
+        }
+    }
+
+
+
+    /** Entry point to load the SMRMaps in memory.
+     *
+     * When this function returns, the maps are fully loaded.
+     *
+     */
+    public void loadMaps() {
+        initializeHeadAndTails();
+
+        long latestRead = -1;
+        long nextRead = 0;
+        while (nextRead <= logTail) {
+            final long start = nextRead;
+            final long stopNotIncluded = Math.min(start + batchReadSize, logTail + 1);
+            nextRead = stopNotIncluded;
+            final Map<Long, ILogData> range = getLogData(start, stopNotIncluded);
+
+            necromancer.execute(() -> {
+                range.forEach((address, logData) -> {
+                    switch (logData.getType()) {
+                        case DATA:
+                            processLogData(logData);
+                            break;
+                        case HOLE:
+                            break;
+                        case TRIMMED:
+                            log.warn("loadMaps[{}, start={}] address is trimmed", address, logHead);
```

Comment:
For now, not. There is currently no way to get the head of the log. Maithem is working on it. If we get the head of the log and we encounter a trim, it might be okay to just go over it. Since we are applying entries sequentially, not applying a TRIM and just go over it is okay.

I will address this (if I have to) in the second pass whith trimm/checkpoint support